python/pyarrow/tests/test_parquet.py

```diff
@@ -2306,6 +2307,30 @@ def test_backwards_compatible_column_metadata_handling(datadir):
     tm.assert_frame_equal(result, expected[['a']].reset_index(drop=True))
 
 
+def test_pickle_dataset(datadir):
+    path = datadir / 'v0.7.1.column-metadata-handling.parquet'
+    dataset = pq.ParquetDataset(path)
+
+    result = pickle.loads(pickle.dumps(dataset))
+    assert dataset == result
+
+    for piece in dataset.pieces:
+        assert piece == pickle.loads(pickle.dumps(piece))
+
+
+def test_cloudpickle_dataset(datadir):
+    cp = pytest.importorskip('cloudpickle')
+
+    path = datadir / 'v0.7.1.column-metadata-handling.parquet'
+    dataset = pq.ParquetDataset(path)
+
+    result = cp.loads(cp.dumps(dataset))
+    assert dataset == result
+
+    for piece in dataset.pieces:
+        assert piece == cp.loads(cp.dumps(piece))
```

Comment:
If we wanted to go further, it might be useful to extend these tests to include ...

```python
dataset.metadata
dataset.metadata.schema
dataset.metadata.schema.column(0)
dataset.metadata.row_group(0)
```

Assuming, of course, that the dataset has a metadata file (and so these things are not `None`)

Reply:
Sure, I can do that. A more interesting question would be the continuous testing of pyarrow and dask.

We have dockerized integration tests for dask and pyarrow, but the visibility currently is pretty bad (which will change soon). May I suggest You to run these test somewhere in your CI as well?

Reply:
> Sure, I can do that. A more interesting question would be the continuous testing of pyarrow and dask.

Indeed

> We have dockerized integration tests for dask and pyarrow, but the visibility currently is pretty bad (which will change soon). May I suggest You to run these test somewhere in your CI as well?

 Sure.  We do have integrated Dask + PyArrow tests in our test suite, but I suspect that they only test against latest release.  Does Arrow publish dev conda builds by any chance?  It would be pretty easy to add this to one of our CI testing builds if so.

Reply:
We already have nightly conda binaries, but not uploaded to anaconda yet. I'll let You know when the binaries are available.

Reply:
Please do.  I'll be glad to submit a PR to dask to integrate them to our CI system.